### PR TITLE
VZ-6247 Uninstall unit tests

### PR DESCRIPTION
**Pull request by pfmackin**
Add uninstall unit tests and fix logic bug in uninstall MC

### CODE REVIEW

File: platform-operator/controllers/verrazzano/uninstall_test.go [hunk 1]
```diff
@@ -267,3 +283,210 @@ func TestReconcileUninstall(t *testing.T) {
 	asserts.Equal("Uninstalled", string(vzcr.Status.Components["fake"].State), "Invalid component state")
 	asserts.NotZero(len(UninstallTrackerMap), "UninstallTrackerMap should have no entries")
 }
+
+func TestUninstallVariations(t *testing.T) {
```
Comment (michael-cico) [inline platform-operator/controllers/verrazzano/uninstall_test.go:287]: Test prologue

**Review by desagar — APPROVED** [review]

File: platform-operator/controllers/verrazzano/uninstall_test.go [hunk 2]
```diff
@@ -267,3 +283,210 @@ func TestReconcileUninstall(t *testing.T) {
...
+			secrets: []corev1.Secret{
+				{ObjectMeta: metav1.ObjectMeta{Name: vzconst.MCAgentSecret, Namespace: vzconst.VerrazzanoSystemNamespace}},
+				{ObjectMeta: metav1.ObjectMeta{Name: vzconst.MCRegistrationSecret, Namespace: vzconst.VerrazzanoSystemNamespace}},
+				{ObjectMeta: metav1.ObjectMeta{Name: "verrazzano-cluster-elasticsearch", Namespace: vzconst.VerrazzanoSystemNamespace}},
+			},
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			initUnitTesing()
```
Comment (michael-cico) [inline platform-operator/controllers/verrazzano/uninstall_test.go:330]: It seems like there's a lot going on in here... is there a way to break it up into multiple test methods?  It's going to be hard to follow and update.

File: platform-operator/controllers/verrazzano/uninstall_test.go [hunk 3]
```diff
@@ -267,3 +283,210 @@ func TestReconcileUninstall(t *testing.T) {
...
+							Type: vzapi.CondInstallComplete,
+						},
+					},
+					Components: func() vzapi.ComponentStatusMap {
+						statusMap := makeVerrazzanoComponentStatusMap()
+						return statusMap
+					}(),
+				},
+			}
+			_ = vzapi.AddToScheme(k8scheme.Scheme)
```
Comment (michael-cico) [inline platform-operator/controllers/verrazzano/uninstall_test.go:376]: maybe do all the scheme updates outside the test flow.

File: platform-operator/controllers/verrazzano/uninstall_test.go [hunk 4]
```diff
@@ -267,3 +283,210 @@ func TestReconcileUninstall(t *testing.T) {
...
+			_ = clustersv1alpha1.AddToScheme(k8scheme.Scheme)
+			_ = vzappclusters.AddToScheme(k8scheme.Scheme)
+
+			// Add core resources
+			cb := fake.NewClientBuilder().WithScheme(k8scheme.Scheme).WithObjects(
+				vzcr,
+				rbac.NewServiceAccount(namespace, name, []string{}, labels),
```
Comment (michael-cico) [inline platform-operator/controllers/verrazzano/uninstall_test.go:383]: these 2 resources are going away once we get rid of the job.

File: platform-operator/controllers/verrazzano/uninstall_test.go [hunk 5]
```diff
@@ -267,3 +283,210 @@ func TestReconcileUninstall(t *testing.T) {
...
+			})
+			defer helm.SetDefaultChartStatusFunction()
+
+			// Sample bom file for version validation functions
+			config.SetDefaultBomFilePath(testBomFilePath)
+			defer func() {
+				config.SetDefaultBomFilePath("")
+			}()
+
+			config.TestProfilesDir = "../../manifests/profiles"
```
Comment (michael-cico) [inline platform-operator/controllers/verrazzano/uninstall_test.go:432]: if we are going to use canned VZ resources, and don't need the profile merging logic, we can probably get rid of this and use the same CR for actual/expected in the fake component context.

Comment (pfmackin) [reply platform-operator/controllers/verrazzano/uninstall_test.go]: test fails without test

File: platform-operator/controllers/verrazzano/uninstall_test.go (hunk 4 above)
Comment (pfmackin) [reply platform-operator/controllers/verrazzano/uninstall_test.go]: they should be removed at time

File: platform-operator/controllers/verrazzano/uninstall_test.go (hunk 3 above)
Comment (pfmackin) [reply platform-operator/controllers/verrazzano/uninstall_test.go:307]: fixed

File: platform-operator/controllers/verrazzano/uninstall_test.go (hunk 2 above)
Comment (pfmackin) [reply platform-operator/controllers/verrazzano/uninstall_test.go]: fixed

File: platform-operator/controllers/verrazzano/uninstall_test.go [hunk 6]
```diff
@@ -267,3 +283,210 @@ func TestReconcileUninstall(t *testing.T) {
...
+
+			c := cb.Build()
+
+			// Sample bom file for version validation functions
+			config.SetDefaultBomFilePath(testBomFilePath)
+			defer func() {
+				config.SetDefaultBomFilePath("")
+			}()
+			// Stub out the call to check the chart status
+			helm.SetChartStatusFunction(func(releaseName string, namespace string) (string, error) {
```
[the remark this answers is not in the document]
Comment (pfmackin) [reply platform-operator/controllers/verrazzano/uninstall_test.go]: removed

File: platform-operator/controllers/verrazzano/uninstall_test.go (hunk 1 above)
Comment (pfmackin) [reply platform-operator/controllers/verrazzano/uninstall_test.go:238]: fixed

File: platform-operator/controllers/verrazzano/uninstall.go [hunk 7]
```diff
@@ -193,20 +196,23 @@ func (r *Reconciler) deleteMCResources(log vzlog.VerrazzanoLogger) error {
...
-		return err
-	}
-	if err := r.deleteSecret(log, vzconst.VerrazzanoSystemNamespace, vzconst.MCAgentSecret); err != nil {
-		return err
+	// Delete secrets on managed cluster.  Don't delete MC agent secret until the end since it tells us this is MC install
+	if managed {
+		if err := r.deleteSecret(log, vzconst.VerrazzanoSystemNamespace, vzconst.MCRegistrationSecret); err != nil {
+			return err
+		}
+		if err := r.deleteSecret(log, vzconst.VerrazzanoSystemNamespace, "verrazzano-cluster-elasticsearch"); err != nil {
```
[the remark this answers is not in the document]
Comment (pfmackin) [reply platform-operator/controllers/verrazzano/uninstall.go]: fixed

**Review by michael-cico — APPROVED** [review]

**Review by bobdonat — DISMISSED** [review]

**Review by michael-cico — APPROVED** [review]

**Review by bobdonat — APPROVED** [review]
